Client/src/StorytellerClient.cpp

```diff
@@ -0,0 +1,233 @@
+{}
+////////////////////
+template<class KeeperChoicePolicy> // = chronolog::RoundRobinKeeperChoice>
+chronolog::StoryWritingHandle<KeeperChoicePolicy>::~StoryWritingHandle()
+{  delete keeperChoicePolicy; }
+
+////////////////////
+template< class KeeperChoicePolicy> 
+void chronolog::StoryWritingHandle<KeeperChoicePolicy>::addRecordingClient(chronolog::KeeperRecordingClient* keeperClient) 
+{
+	storyKeepers.push_back(keeperClient);  
+}
+///////////////////
+template< class KeeperChoicePolicy> 
+void chronolog::StoryWritingHandle<KeeperChoicePolicy>::removeRecordingClient(chronolog::KeeperIdCard const& keeper_id_card) 
+{
+	// this should only be called when the ChronoKeeper process unexpectedly exits 
+	// so it's ok to use rather inefficient vector iteration....
+	for( auto iter = storyKeepers.begin(); iter != storyKeepers.end(); ++iter)
+	{
+	 // if( (*iter)->keeperId == keeper_id_card)
+	 // {  storyKeepers.erase(iter); break; }
+	}
+}
+//////////////////
+template< class KeeperChoicePolicy> 
+int chronolog::StoryWritingHandle<KeeperChoicePolicy>::log_event( std::string const& event_record)
+{
+
+	chronolog::LogEvent log_event(storyId, 
+			theClient.getTimestamp(), 
+			theClient.getClientId(), 
+			theClient.get_event_index(), 
+			event_record);
+
+	auto keeperRecordingClient = keeperChoicePolicy->chooseKeeper(storyKeepers, log_event.time());
+
+	if(nullptr == keeperRecordingClient)   //very unlikely... 
+	{  return 0; }
+
+	// INNA: make send event returm 0 in case of tl RPC failure .... 
+	keeperRecordingClient->send_event_msg( log_event);
+
+return 1;
+}
+/////////////////////
+
+template< class KeeperChoicePolicy> 
+int chronolog::StoryWritingHandle<KeeperChoicePolicy>::log_event( size_t ,  void*)
+{
+  return 0;  // not implemented yet ; to be implemented with tl bulk transfer ... 
+}
+
+chronolog::StorytellerClient::~StorytellerClient()
+{
+ //INNA: TODO not implemented yet
+   std::cout<<"StorytellerClient::~StorytellerClient()"<<std::endl;
+   //release acquired stories
+
+   // stop & delete keeperRecordingClients
+   std::lock_guard<std::mutex> lock(recordingClientMapMutex);
+   for ( auto keeper_client : recordingClientMap)
+   {  
+	delete keeper_client.second; 
+   }
+   recordingClientMap.clear();
+}
+
+int chronolog::StorytellerClient::get_event_index()
+{
+	// we only aqcuire mutex in the rare case when 
+	// the atomic index has reached INT_MAX value 
+	// otherwise proceed lock-free
+	if( (atomic_index == INT_MAX) ) 
+	{
+	   std::lock_guard<std::mutex> lock(recordingClientMapMutex);
+	   //recheck when mutex is acquired, only one thread changes the value
+	   if(atomic_index == INT_MAX) 
+	   {    atomic_index = 0; }
+	}
+return ++atomic_index;
+}
+////////////////
+
+
+int chronolog::StorytellerClient::addKeeperRecordingClient( chronolog::KeeperIdCard const& keeper_id_card)
+{
+	   std::lock_guard<std::mutex> lock(recordingClientMapMutex);
+
+           std::string service_na_string("ofi+sockets://");
```

Comment:
I am assuming protocol string from confmanager will be used later here.

Reply:
Yes, the protocol string will have to be passed in, not hardcoded . though I'm not sure we can assume that all the nodes will have the same fabrics available.